### DISCUSSION TITLE
Receiving stale values for conditional dependents in atom read

### DISCUSSION THREAD

**Discussion by probeiuscorp** [pmndrs/jotai#2532] [Bug report] [answered]
In certain circumstances atoms with conditional dependents may receive stale values for those dependents.

I was able to reproduce the issue (not easily!) and have made a [test case](https://github.com/probeiuscorp/jotai/commit/b38f9003607c356aa9935267c3a79009ba3f3ec5).

A few notes:
 * Moving `const filtered = get(filteredAtom)` up before the `if` in `stageAtom`'s read avoids the bug
   * This is how I've mitigated this bug in my own code
 * Subscribing to both `stageAtom` and `filteredAtom` is necessary to reproduce the bug

This is a regression since 59e04ab. In my testing store2 passes.

**Comment by dai-shi** [comment]
Nice catch.

> In my testing store2 passes.

Do you mean, `store.ts` fails while `store2.ts` passes? That's great to know.
Please open a PR with this condition: `it.skipIf(!import.meta.env?.USE_STORE2)(...)`

Example: https://github.com/pmndrs/jotai/pull/2533/files#diff-d32cc71c13ecc7fc7879c5f6bd5a11066daa3e2b351fd23668fcf8d9438911cbR29

**Reply by probeiuscorp** [reply]
I didn't try the test case with those exact conditions, so I can't say for sure, but I built `jotai/experimental` and it worked for my application code like v2.6.3 did. I'll be opening a PR soon.

**Reply by dai-shi** [reply]
Sounds good.

**Reply by probeiuscorp** [reply, marked as answer]
PR is in!

**Reply by dai-shi** [reply]
> I didn't try the test case with those exact conditions, so I can't say for sure, but I built `jotai/experimental` and it worked for my application code like v2.6.3 did. I'll be opening a PR soon.

https://github.com/pmndrs/jotai/pull/2534#issuecomment-2084357797 Can you confirm if it works with `jotai/experimental` again? The test doesn't seem to pass. If it's the case, our approach might change.

**Reply by probeiuscorp** [reply]
Since that comment I have forgotten what I did to use `jotai/experimental`. Could you remind me?

**Reply by dai-shi** [reply]
Something like this:
```jsx
import { createStore } from 'jotai/experimental';
import { Provider } from 'jotai/react';

const store = createStore();

const MyProvider = ({ children }) => <Provider store={store}>{children}</Provider>;
```

**Reply by probeiuscorp** [reply]
In hindsight I definitely should've tried that. Yes, for me `jotai/experimental` did not pass the test either. Sorry for the wrong info.
